lib/hexpm_web/templates/docs/self_hosting.html.md

```diff
@@ -0,0 +1,360 @@
+
+```
+$ mix hex.package fetch decimal 2.0.0
+decimal v2.0.0 downloaded to decimal-2.0.0.tar
+$ cp decimal-2.0.0.tar public/tarballs/
+$ mix hex.registry build public --name=acme --private-key=private_key.pem
+* creating public/packages/decimal
+* updating public/names
+* updating public/versions
+```
+
+Now let's test our repository. We can use the built-in server that ships with Erlang/OTP to serve the `public` directory:
+
+```
+$ erl -s inets -eval 'inets:start(httpd,[{port,8000},{server_name,"localhost"},{server_root,"."},{document_root,"public"}]).'
+```
+
+And let's now add the repository and try fetching the package that we just published:
+
+```
+$ mix hex.repo add acme http://localhost:8000 --public-key=public/public_key
+$ mix hex.package fetch decimal 2.0.0 --repo=acme
+decimal v2.0.0 downloaded to decimal-2.0.0.tar
+```
+
+If everything went well you should see that the package was downloaded from your local server!
+
+In the next sections we'll cover how we can deploy our registry to production.
+
+### Deploying to S3
+
+Deploying to Amazon S3 (or similar cloud services) is probably the easiest way to have a reliable Hex repository.
+
+If you already have an S3 bucket, use e.g. [AWS CLI](https://aws.amazon.com/cli/) to sync the contents of the `public/` directory
+like this:
+
+```
+$ aws s3 sync public s3://my-bucket
+```
+
+**Warning:** Remember to sync only the public directory and not `private_key.pem`! And if you do want to sync your private key, remember to set appropriate bucket policy so it isn't accidentally exposed.
+
+Your repository should now be available under an URL like: `https://<bucket>.s3.<region>.amazonaws.com` or however you configured your bucket.
+
+If you don't yet have a bucket, create one! By default, files stored on S3 are not publicly accessible. You can enable public access by setting the following bucket policy in your bucket's properties:
+
+```json
+{
+  "Version": "2008-10-17",
+  "Statement": [
+    {
+      "Sid": "AllowPublicRead",
+      "Effect": "Allow",
+      "Principal": {
+        "AWS": "*"
+      },
+      "Action": "s3:GetObject",
+      "Resource": "arn:aws:s3:::my-bucket/*"
+    }
+  ]
+}
+```
+
+You may also consider adding an IAM policy for the user accessing the bucket:
+
+```json
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Sid": "VisualEditor0",
+      "Effect": "Allow",
+      "Action": [
+        "s3:PutObject",
+        "s3:GetObject",
+        "s3:ListBucket",
+        "s3:DeleteObject"
+      ],
+      "Resource": [
+        "arn:aws:s3:::my-bucket",
+        "arn:aws:s3:::my-bucket/*"
+      ]
+    }
+  ]
+}
+```
+
+See [Amazon S3 docs](https://docs.aws.amazon.com/s3/index.html) for more information and remember to customize bucket/IAM policies in a way that makes sense for your deployment.
+
+### Deploying with Plug.Cowboy & Docker
```

Comment:
Maybe this section would be helped with an example github repository that contains the boilerplate and we can trim down on some of the code sections?

Reply:
Yeah, I was thinking about that too. With the excerpt from mix.exs, two module files, and a full Dockerfile, I thought it's borderline ok to just show all the code instead of linking elsewhere but happy to add a new repo. I think the upside is folks will be able to just clone it but the downside is we'd have yet another thing to maintain and keep in sync with the docs if we ever change things.

Do you have particular code sections you'd like to trim? How? Here's what I was thinking:

I could just show the `children = [...]` part of the supervision tree (I included the full module, because that `require Logger + Logger.info("starting ...")` is really helpful), and trim down the Dockerfile. I thought showing `MyApp.Plug` and `config/runtime.exs` in entirety was important as that's the meat of the project. (It'd be great if there was an easier way to configure Static & BasicAuth plugs at _runtime_ cause that would definitely remove some code wink wink.)

With these small changes in place, would you still go ahead with an external repo? If so, how does hexpm/example_repo sound? I'm happy to put it on my account instead, totally fine either way.

Reply:
Since you trimmed some of the code I think we can keep it in docs only for now.